merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/CheckpointSimulationDriver.java

```diff
@@ -0,0 +1,412 @@
+      Duration endsAt,
+      Map<ActivityDirectiveId, ActivityDirective> activityDirectives,
+      SimulationEngine simulationEngine,
+      LiveCells cells,
+      SlabList<TemporalEventSource.TimePoint> timePoints,
+      Topic<ActivityDirectiveId> activityTopic,
+      MissionModel<?> missionModel
+  ) {
+    public void freeze() {
+      cells.freeze();
+      timePoints.freeze();
+      simulationEngine.close();
+    }
+
+    public static CachedSimulationEngine empty(final MissionModel<?> missionModel) {
+      final SimulationEngine engine = new SimulationEngine();
+      final TemporalEventSource timeline = new TemporalEventSource();
+      final LiveCells cells = new LiveCells(timeline, missionModel.getInitialCells());
+
+      // Begin tracking all resources.
+      for (final var entry : missionModel.getResources().entrySet()) {
+        final var name = entry.getKey();
+        final var resource = entry.getValue();
+
+        engine.trackResource(name, resource, Duration.ZERO);
+      }
+
+      {
+        // Start daemon task(s) immediately, before anything else happens.
+        engine.scheduleTask(Duration.ZERO, missionModel.getDaemon());
+        {
+          final var batch = engine.extractNextJobs(Duration.MAX_VALUE);
+          final var commit = engine.performJobs(batch.jobs(), cells, Duration.ZERO, Duration.MAX_VALUE);
+          timeline.add(commit.getKey());
+        }
+      }
+      final var emptyCachedEngine = new CachedSimulationEngine(
+          Duration.MIN_VALUE,
+          Map.of(),
+          engine,
+          cells,
+          timeline.points(),
+          new Topic<>(),
+          missionModel
+      );
+      return emptyCachedEngine;
+    }
+  }
+
+  /**
+   * Selects the best cached engine for simulating a given plan.
+   * @param schedule the schedule/plan
+   * @param cachedEngines a list of cached engines
+   * @return the best cached engine as well as the map of corresponding activity ids for this engine
+   */
+  public static Optional<Pair<CachedSimulationEngine, Map<ActivityDirectiveId, ActivityDirectiveId>>> bestCachedEngine(
+      final Map<ActivityDirectiveId, ActivityDirective> schedule,
+      final List<CachedSimulationEngine> cachedEngines) {
+    Optional<CachedSimulationEngine> bestCandidate = Optional.empty();
+    final Map<ActivityDirectiveId, ActivityDirectiveId> correspondenceMap = new HashMap<>();
+    for (final var cachedEngine : cachedEngines) {
+      if (bestCandidate.isPresent() && cachedEngine.endsAt().noLongerThan(bestCandidate.get().endsAt()))
+        continue;
+
+      final var activityDirectivesInCache = new HashMap<>(cachedEngine.activityDirectives());
+      // Find the invalidation time
+      var invalidationTime = Duration.MAX_VALUE;
+      final var scheduledActivities = new HashMap<>(schedule);
+      for (final var activity : scheduledActivities.entrySet()) {
+        if (activityDirectivesInCache.values().contains(activity.getValue())) {
```

Comment:
Minor, non-blocking, simplify:
```suggestion
        if (activityDirectivesInCache.containsValue(activity.getValue())) {
```